steward/src/cork.rs

```diff
@@ -0,0 +1,107 @@
+use crate::{
+    cellars::{self, uniswapv3, CellarId},
+    error::{Error, ErrorKind},
+    prelude::APP,
+    somm_send,
+};
+use abscissa_core::{
+    tracing::log::{debug, error},
+    Application,
+};
+use deep_space::{Coin, Contact};
+use gravity_bridge::gravity_proto::cosmos_sdk_proto::cosmos::base::abci::v1beta1::TxResponse;
+use somm_proto::cork::Cork;
+use std::time::Duration;
+use steward_proto::steward::{
+    self,
+    submit_request::ContractCallData::{self, *},
+    SubmitRequest, SubmitResponse,
+};
+use tonic::async_trait;
+
+pub struct CorkHandler;
+
+#[async_trait]
+impl steward::contract_call_server::ContractCall for CorkHandler {
+    async fn submit(
+        &self,
+        request: tonic::Request<SubmitRequest>,
+    ) -> Result<tonic::Response<SubmitResponse>, tonic::Status> {
+        debug!("received contract call request");
+        tokio::spawn(async move {
+            let request = request.get_ref();
+            let cork = match build_cork(request).await {
+                Ok(c) => c,
+                Err(err) => {
+                    error!("{}", err);
+                    return;
```

Comment:
In these error cases, shouldn't we return a relevant `tonic::Status` indicating an error condition? Can we propagate a value like this out of the `tokio` thread?

Reply:
It looks like the cases where `build_cork` errors are when
1. The [cellar_id is improperly formatted](https://github.com/PeggyJV/steward/blob/da905a21c2ec8f892cc36a048184907510855f28/steward/src/cork.rs#L53) (address is wrong length, there is no ':', etc.)
2. The [contract call data is missing](https://github.com/PeggyJV/steward/blob/da905a21c2ec8f892cc36a048184907510855f28/steward/src/cork.rs#L55-L58)
3. The [`cellar address is not whitelisted in the steward config](https://github.com/PeggyJV/steward/blob/da905a21c2ec8f892cc36a048184907510855f28/steward/src/cork.rs#L75)

I can see all of these being covered by an `InvalidArguments` code wrapped in the `tonic::Status`. Do you think that is appropriate?

Reply:
The first two are cases of invalid arguments, the third I'd say is an authorization failure.